maltoolbox/attackgraph/analyzers/apriori.py

```diff
@@ -173,3 +179,54 @@ def prune_unviable_and_unnecessary_nodes(graph: AttackGraph) -> None:
         if (node.type == 'or' or node.type == 'and') and \
             (not node.is_viable or not node.is_necessary):
             graph.remove_node(node)
+
+
+def propagate_viability_from_unviable_node(
+        unviable_node: AttackGraphNode,
+        attack_steps_made_unviable: Optional[list[AttackGraphNode]] = None
+    ) -> list[AttackGraphNode]:
+    """
+    Update viability of nodes affected by newly enabled defense
+    `unviable_node` in the graph and return any attack steps
+    that are no longer viable because of it.
+
+    Propagate recursively via children as long as changes occur.
+
+    Arguments:
+    unviable_node               - the node to propagate viability from
+    attack_steps_made_unviable  - list of the attack steps that have been
+                                  made unviable by a defense enabled in the
+                                  current step. Builds up recursively.
+    """
```

Comment:
I don't like the fact that it has the result both as a parameter and the return value. I would have it only in the return and build it up via recursion.